modules/core/include/opencv2/core/cuda.hpp

```diff
@@ -656,6 +656,19 @@ class CV_EXPORTS_W Stream
     //! creates a new asynchronous stream with custom allocator
     CV_WRAP Stream(const Ptr<GpuMat::Allocator>& allocator);
 
+    /** @brief creates a new Stream from a cudaStream_t and takes ownership of that CUDA stream
+
+    @note The parameter must be a cudaStream_t. The parameter is void* to align
+    with cudaPtr(). The ownership of the cudaStream_t will move to this Stream.
+    The destructor of this Stream will cudaStreamDestroy(cudaStream).
+    @code
+        cudaStream_t cudaStream;
+        cudaStreamCreateWithFlags(&cudaStream, cudaStreamNonBlocking);
+        cv::cuda::Stream cvStream(std::move((void*)cudaStream));
+    @endcode
+     */
+    Stream(void*&& cudaStream);
```

Comment:
It make sense to avoid "void*" ctors in public API (because there is no strong type check and many implicit conversions).

Consider using:
```
static Stream fromCudaStream(void*&& cudaStream);
```

Usage:

```.diff
-cv::cuda::Stream cvStream(std::move((void*)cudaStream));
+auto cvStream = cv::cuda::Stream::fromCudaStream(std::move((void*)cudaStream));
```

Reply:
The ask is a constructor for the reasons that I wrote in the original issue https://github.com/opencv/opencv/issues/19285. Please reference there for details including the unusable existing function `cv::cuda::StreamAccessor::wrapStream()`. 

A constructor can provide the instantiated object itself or a pointer to an instantiated object. They are not the same. A pointer is raw and unsafe to use without further wrapping it with unique_ptr or shared_ptr. The object itself is scoped, RAII, etc. Technically both are valid. But only a constructor can create an object itself into an lvalue.

```c++
// safe, destructs when out of scope
cv::cuda::Stream stream_obj(std::move((void *)my_cudaStream_t));

// dangerous raw pointer, does not destruct itself, often multiple memory bugs by app programmers
cv::cuda::Stream* stream_ptr = cv::cuda::Stream::fromCudaStream(std::move((void *)my_cudaStream_t));

// safer, valid, more work
std::unique_ptr<cv::cuda::Stream> stream_ptr(cv::cuda::Stream::fromCudaStream(std::move((void *)my_cudaStream_t)));

// safe, simple, and requires a constructor
struct threadTLSData{
    cv::cuda::Stream cvStream;
    int otherThings;
};
return {
    cv::cuda::Stream(std::move((void *)_stream)),
    42
};

// without a constructor have to wrap it all and requires C++11 compiler correctly supports copy elision
struct threadTLSData{
    std::unique_ptr<cv::cuda::Stream> cvStream; // share_ptr isn't as strict with elision, but has more overhead when we don't need it here
    int otherThings;
};
return {
    // also can not use std::make_unique or std::make_shared because they both require a constructor
    // also make_unique is only in C++14 and newer yet OpenCV wants to target C++11
    // this direct unique_ptr constructor method is often unsafe in C++11 due to parameter evaluation order
    // with exceptions. For safety, need make_unique in C++14 -or- use C++17 standard that fixed this
    // https://www.learncpp.com/cpp-tutorial/stdunique_ptr/#urvanov-syntax-highlighter-600141f5f0789724071531
    std::unique_ptr<cv::cuda::Stream>(cv::cuda::Stream::fromCudaStream(std::move((void *)my_cudaStream_t))),
    42
};
```

OpenCV is often not strong type checked. As I discuss in the issue #19285, this is a development choice by the core team. They did not want to include headers (e.g. CUDA, OpenCL) and fix issues at compile time, so they wrote generic parameters/returns like `void*` and throw asserts at runtime. For example: `cv::ocl::Device(void*)`, `void* cv::cuda::Stream::cudaPtr()`, `cv::cuda::Stream::Stream()` that throws, or the whole auxiliary namespace `cv::cuda::StreamAccessor`.

I would have taken a different approach with strong type safety and fix at compile time.  However since the core team choose the generic `void*` and assert approach, I suggest we continue that approach for consistency plus the above benefits of having a constructor. I've coded it like that in the PR.

### Options

#### Constructor with `Stream::Stream(void*&& cudaStream)`

See original post and PR

#### Static factory `static Stream::fromCudaStream(void*&& cudaStream)`

I can write a lesser feature with a static factory like `static cv::cuda::Stream::fromCudaStream(void*&& cudaStream)`. It will have the limitations and more difficult usage that I describe above in words and pseudocode.

#### Use type-safety

Make a breaking change to the `cv::cuda` module to end the use of `void*` on public APIs. This requires the inclusion some cuda headers so that types like `cudaStream_t` can be used in APIs like this new constructor, existing `Stream::cudaPtr()`, etc.

#### Gated availability of type-safe constructor

It should be easy to use `#ifdef HAVE_CUDA` to test for availability of CUDA, optionally include CUDA headers and declare a type-safe constructor. This would give us a constructor, type-safe, and not required CUDA headers to be loaded unless OpenCV is being built with cuda. For example pseudocode

```
#ifdef HAVE_CUDA
    #include<cuda.h>
    cv::cuda::Stream::Stream(cudaStream_t&& cudaStream);
#endif
```

Reply:
> cv::ocl::Device(void*)

Exactly in this area there are several issues with "void*" type in constructor.

> unique_ptr

Doesn't make sense, because this class is smart ptr itself.
Anyway you can always try to add this "safe" factory:

```
static std::unique_ptr<Stream> makeUniqueFromCudaStream(void*&& cudaStream);
```

> copy elision

Try to add move ctor.

---

For other proposals there is ABI compatibility (between configurations) with BUILD_CUDA_STUBS CMake option.

Reply:
I believe there is a possibility here. We just haven't found something that you will accept into the codebase and that meets the needs of users like me. What else should we explore?

> Exactly in this area there are several issues with "void*" type in constructor.

Help me understand. Are you writing as a core team member....that the core team *does not* want to have `void*` in public opencv APIs? I ask because `void* Stream::cudaPtr()` and three others was added 11 months ago on 26 February 2020. These `void*` uses are not old...they are new. Someone in the core team allowed it, here is the commit https://github.com/opencv/opencv/commit/53c0189ed7efe52ad65c38cac04181155e320ca8

It is difficult for contributors to understand how to contribute when there is conflicting information. You are telling me one thing (don't use void*) and only 11 months ago I see it was allowed and merged (with more `void*` use if we look older). My intention is not to complain or make the core team wrong. Instead, I want to contribute in a positive way. I need clear consistent strategy and implementation rules. If I don't have that, it is difficult for me to know how to reverse-engineer and write new code. 🤪

> unique_ptr

This is a important need. Static factories return *pointers* to objects; it returns `cuda::Stream *`. This is different than constructors. A constructor returns `cuda::Stream`. If we take the static factory approach, then it returns a pointer...therefore I want to make it safe. No memory leaks, ref count errors, double freeing, ownership, etc! 🔨 These are real pointer problems that happen in opencv (I've found and fixed some) and in apps that use opencv as a library. We are human, we make mistakes. I strongly prefer designs that prevent these mistakes.  C++11 and newer has good mechanisms to prevent them.

When adding a constructor, we get all the benefits of RAII and scope that I write above. If instead, we have a static factory that returns a pointer, then how do we make it safe? Yes, it could return a unique_ptr instead. That aligns with C++ core guideline 26 https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f26-use-a-unique_ptrt-to-transfer-ownership-where-a-pointer-is-needed

However, that doesn't align with the typical use of static factories like discussed in guideline https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c50-use-a-factory-function-if-you-need-virtual-behavior-during-initialization. Usually static factories are used for one of two reasons:

1. To return a pointer to a class object from a pool of reusable existing objects, or to return a singleton object.
2. To return a pointer to a subclass based on the parameters passed. It does not return a pointer to itself (the base object)

A static factory like below could technically work, but it "feels" strange, lesser, and not a good precedent to set.
```c+
static std::unique_ptr<cv::cuda::Stream> cv::cuda::Stream::makeUniqueFromCudaStream(void*&& cudaStream);
```

This feeling is so strong, I would suggest not doing it. Instead, overloading a function in that `cv::cuda::StreamAccessor` namespace where it can be strongly typed. Pseudocode...

```c++
Stream cv::cuda::StreamAccessor::wrapStream(cudaStream_t&& stream);
```

It is consistent with the already existing wrapStream() function. But, it is not consistent with the already existing `void*` use. Personally, I probably wouldn't use a function like this. It can't be used to initialize a member field; at best a field would be default constructed and then copy assigned from this function. I want a constructor for safety and simplicity.

> Try to add move ctor.

That is my original post and PR. It is a move constructor.

Reply:
> "void*" type in constructor.
> does not want to have void* in public opencv APIs

The first only.

> Static factories return pointers to objects; it returns cuda::Stream *

This doesn't follow RAII.

---

> StreamAccessor

I believe this is the right way.
There is already some related strong-typed [code](https://github.com/opencv/opencv/blame/63bb2abadab875fc648a572faccafee134f06fc8/modules/core/src/cuda_stream.cpp#L563-L566) (added in C++98 epoch, so there is no move semantic).

Reply:
/cc @nglee @cudawarped @tomoaki0705 Any through about this?

Reply:
To the best of my understanding, `cudaStream_t` is itself a pointer.
```C++
// driver_types.h
typedef __device_builtin__ struct CUstream_st *cudaStream_t;
```

So I think it's not clear why we would need a constructor that takes ownership of a `cudaStream_t`. We can create a `cv::cuda::Stream` object and get the managed `cudaStream_t` object with `cv::cuda::StreamAccessor::getStream(const Stream& stream)`.

Reply:
I think the constructor that takes ownership of a `cudaStream_t` would need to check whether the `cudaStream_t` passed as an argument is already initialized with `cudaStreamCreate()` or not. I can't find an API in CUDA that does that. So I think it is hard to implement one.

I'm not sure in which situation we would need such a constructor.

Also, `cudaStream_t` is itself a pointer, so I think there are not many benefits from `std::move`-ing it.
```C++
// driver_types.h
typedef __device_builtin__ struct CUstream_st *cudaStream_t;
```

Reply:
Move is the C++11 and newer semantic to declare and force a transfer of ownership. Without it, ownership is ambigious -- often duplicated, and leading to leaks and/or double-deletes.
After the move into the constructor I proposed...the previously owning variable has no data and no memory. I can't be used further in code. These are the good things that C++11 and newer has to help coders write better code. That is the value: clear ownership....less bugs...cleaner code.

I previously looked myself, CUDA doesn't have an API to validate a cudaStream_t. However, this is something I would not expect OpenCV to do. OpenCV doesn't validate the `data` pointer for cv::mat, cv::umat or any of many many other classes that have constructors accepting pointers.

Reply:
I wanted to say that move semantics is not that useful for primitive types such as pointers.

https://stackoverflow.com/questions/14679605/do-built-in-types-have-move-semantics

I think just declaring the constructor as:
```C++
Stream(void* cudaStream);
```
and calling it as the following seems enough.
```C++
cudaStream_t cudaStream;
cudaStreamCreateWithFlags(&cudaStream, cudaStreamNonBlocking);
cv::cuda::Stream cvStream((void*)cudaStream));
```

However, speaking of other OpenCV constructors, those that accept pointers do not take ownership. I'm not sure if we need one for `Stream`.

Reply:
Snap! I have an alternative.

Thanks @nglee you gave the hint at the std::move doesn't actually do what I though it did. More research and I now see that std::move is very much only a cast and that it is the responsibility of the constructor/assignment operator to manually change memory. For example STL leaves the var "placed in a valid but unspecified state".

I could change my PR to set the `cudaStream_t` to null. Unfortunately, a null cudaStream_t (the pointer itself rather than to which it points) means the default CUDA stream. This would prevent double cudaStreamDestroy() because the destroy api only works on async streams and would likely return an error or silent ignore (I could test it). Is this any better? Ehhh...marginally. It does transfer ownership of `cudaStream_t`. But that means the previously owning var now has a value of null which a programmer could continue to use in errant code to do things without realizing it is now with the default CUDA stream.

The legacy of the `cudaStream_t` might be the death of this. I do not see any CUDA define for an invalid stream. (They have them for devices and cpus...but not streams). I have no value that I can set the previously owning var so that it can not be misused.

My main use case is to create a Stream with a specific type of cudaStream_t. I need unique per-thread non-default async streams. These can be created with `cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking)`. OpenCV aready has code with this approach at `cv::dnn::cuda4dnn::csl::UniqueStream()`

The prototype for that nvidia cuda API is: `cudaStreamCreateWithFlags(cudaStream_t *pStream, unsigned int flags)`

### Alternative

Have a Stream constructor like this

```c++
CV_WRAP cv::cuda::Stream::Stream(unsigned int cudaFlags);

// example use
cv::cuda::Stream mystream(cudaStreamNonBlocking);
```

Benefits are:

* It is a constructor with all its benefits
* No CUDA headers are needed in the public opencv API
* Ownership is always contained within the Stream()
* It can be used from Python

Reply:
The alternative seems good to me. 👍

Reply:
I've updated branch and force pushed. It should arrive soon.